---
title: "scGate: Purification of cell types from complex single-cell RNA-seq data. Basic examples."
author: "M. Andreatta, A. Berenstein and S. Carmona"
date: "29/09/2021"
output:
  rmdformats::readthedown:
    self-contained: true
    highlight: haddock
    thumbnails: false
    css: styles.css
knit: (function(input_file, encoding) {
  out_dir <- 'docs';
  rmarkdown::render(input_file, encoding=encoding,
 output_file=file.path(dirname(input_file), out_dir, 'scGate.basic.usage.html'))})
---

## Setup environment

```{r, message=F, warning=F,results=F}
renv::activate()
renv::restore()
library(ggplot2)
library(dplyr)
library(patchwork)
remotes::install_github("carmonalab/scGate",ref = "dev")
library(scGate)
```


```{r setup, echo=FALSE, message=F, warning=F, results=F}
library(knitr)
library(rmdformats)
library(formatR)


## Global options
options(max.print="75")
opts_chunk$set(echo=TRUE,
	             cache=TRUE,
               cache.lazy=FALSE,
               prompt=FALSE,
               tidy=TRUE,
               comment=NA,
               message=FALSE,
               warning=FALSE,
               dev='png')
opts_knit$set(width=75)

```


## Load demo datasets 
```{r}
testing.datasets <- scGate:::get_testing_data(version = 'hsa.latest')
dir.create('./plots')
```

Let's play with a PBMC dataset ("Satija") from the publication XXX
```{r,collapse =T}
obj <- testing.datasets[["Satija"]]
DimPlot(obj, label = T, repel = T, group.by = "celltype.l1") + theme(legend.position = "none", aspect.ratio = 1) 
ggsave("plots/umap.satija.celltype.3.pdf",width = 3, height = 3)
```

## Creating a simple gating model

Now let's setup a simple scGate gating model to purify a population of interest, for instance, B cells.
B cells are usually identified by the expression of CD20, encoded by MS4A1

```{r}
# model initialization
my_scGate_model <- edit_model()  

# add a signature/set of marker genes
my_scGate_model <- edit_model(model = my_scGate_model, name = "Bcell", signature = c("MS4A1"))
my_scGate_model
```

## Purify a cell population of interest with scGate
```{r, collapse=T}
# run scGate
obj <- scGate(data = obj, model = my_scGate_model, assay = DefaultAssay(obj))
DimPlot(obj, cols = c(list(Impure = "gray", Pure = "green"))) + theme(aspect.ratio = 1) 
ggsave("plots/umap.satija.Bcell.pdf",width = 4, height = 4)
```

Because in this demo example we know the ground truth (at least, as defined by the authors), let's evaluate scGate predictive performance in terms of precision/positive predictive value, recall/sensitivity, and accuracy using Matthews Correlation Coefficient (MCC)
```{r}
scGate::performance.metrics(grepl("^B ",obj$cell_type),obj$is.pure=="Pure")
```
With this very simple model, >99% (Recall) are isolated with a >99% purity (Precision), and and overall accuracy/MCC >99%


In another example, let's gate plasmacytoid dendritic cells (pDCs), which can be isolated using marker LILRA4 (eg https://pubmed.ncbi.nlm.nih.gov/30395816/)
```{r}
my_scGate_model <- edit_model()  # model initizlization
my_scGate_model <- edit_model(model = my_scGate_model, name = "pDC", signature = c("LILRA4"))  # add one positive signature

# run the model
obj <- scGate(data = obj, model = my_scGate_model, assay = DefaultAssay(obj))
DimPlot(obj, cols = c(list(Impure = "gray", Pure = "green"))) + theme(aspect.ratio = 1)

```

```{r}
scGate::performance.metrics(obj$cell_type=="pDC",obj$is.pure=="Pure")
```


## Gating models with positive and negative markers

Another example for gating NKs, which express KLRD1. However, KLRD1 can also be expreseed by some T cell subsets, so we can filter these out by including "negative" T cell markers such as CD3D
```{r}
my_scGate_model <- edit_model()  
my_scGate_model <- edit_model(model = my_scGate_model, name = "NK", signature = c("KLRD1","CD3D-"))  

# run the model
obj <- scGate(data = obj, model = my_scGate_model, assay = DefaultAssay(obj))
DimPlot(obj, cols = c(list(Impure = "gray", Pure = "green"))) + theme(aspect.ratio = 1) 
ggsave("plots/umap.satija.NK.pdf",width = 4, height = 4)

```

Let's evaluate performance
```{r}
scGate::performance.metrics(grepl("^NK",obj$cell_type),obj$is.pure=="Pure")
```

With this simple model, >97% of NKs (recall) are isolated with a 96% purity (Precision), and overall accuracy/MCC of 96%


## Hierarchical gating models

Here we have been using a dataset derived from blood (PBMC). When working with more complex tissues, such as tumors, we might need to apply more stringent criteria. That's why scGate allows to purify by steps, using a hierarchical gating model.

Let's explore the whole tumor dataset by Jerby-Arnon et al (CITE)
```{r,collapse =T}
obj <- testing.datasets[["JerbyArnon"]]
DimPlot(obj, label = T, repel = T, group.by = "cell_type") + theme(legend.position = "none", aspect.ratio = 1) 
ggsave("plots/umap.jerby.celltype.3.pdf",width = 3, height = 3)
```
Here we have non-immune populations such as malignant/cancer cells (Mal), Endothelial cells (Endo) and cancer-associated fibroblasts (CAF).
We can try to purify a broadly defined cell population, for instance, all immune cells, using pan-immune cell marker CD45 encoded by PTPRC:

```{r}
my_scGate_model <- edit_model()  # model initizlization
my_scGate_model <- edit_model(model = my_scGate_model, name = "immune", signature = c("PTPRC")) 
obj <- scGate(data = obj, model = my_scGate_model, assay = DefaultAssay(obj))
DimPlot(obj, cols = c(list(Impure = "gray", Pure = "green"))) + theme(aspect.ratio = 1) 
ggsave("plots/umap.jerby.immune.pdf",width = 4, height = 4)
```

```{r}
scGate::performance.metrics(obj$cell_type %in% c("B.cell","NK","T.CD4","T.CD8","T.cell","Macrophage"),obj$is.pure=="Pure")
```

Now let's try to recover macrophages (eg using common markers CD68 and FCGR1A). Instead of purifying macrophages directly from the whole tissue, we can try to do it in two steps (hierarchical strategy): i) isolate immune cells as in the previous example, ii) isolate macrophages from immune cells. 
Hierarchical gating models can be specified in scGate using parameter "level",  as follows:

```{r}
my_scGate_model <- edit_model()  # model initizlization
my_scGate_model <- edit_model(model = my_scGate_model, name = "immune", signature = c("PTPRC"), level = 1)  # add positive signature at first step
my_scGate_model <- edit_model(model = my_scGate_model, name = "macrophage", signature = c("CD68","FCGR1A"), level = 2)  # add positive signature at second step
obj <- scGate(data = obj, model = my_scGate_model, assay = DefaultAssay(obj))
DimPlot(obj, cols = c(list(Impure = "gray", Pure = "green"))) + theme(aspect.ratio = 1) 
ggsave("plots/umap.jerby.macro.pdf",width = 4, height = 4)
```

```{r}
scGate::performance.metrics(obj$cell_type %in% c("Macrophage"),obj$is.pure=="Pure")
```

Here we isolated 92% of macrophages with a 91% purity (according to annotation by Jerby-Arnon et al). We could easily improve PREC and REC by adding more positive and negative markers, respectively (e.g. removing lymphocytes)


We can see how the two-level gating model worked in each step using scGate's plot_levels.
We see on the left the immune cell purification step, and on the right, the macrophage purifications step.
```{r}
wrap_plots(plot_levels(obj))
```

## Editing models
scGate models of arbitrary complexity can be easily constructed to achieve high purification performance.
These gating models can be written using the edit_model(), and manually edited using eg R's fix().
You can also export your basic model created with edit_model() as tabulated text and edit on Excel
```{r, eval=F}
fix(my_scGate_model) # edit locally in R
write.table(my_scGate_model,"test.tsv",sep="\t") # export and then edit on Excel
my_scGate_model_edited <- scGate::load_scGate_model("test.tsv") # tabulated-text model can loaded
```


## Fast evaluation of model performance

We provide a built-in function `test_my_model` to automatically evaluate the performance of a gating model using 3 pre-annotated testing datasets:

```{r, fig.width=8,fig.height=12,collapse = T, message = F}
my_scGate_model <- edit_model()  
my_scGate_model <- edit_model(model = my_scGate_model, name = "Bcell", signature = c("MS4A1"))
panBcell.performance <- test_my_model(my_scGate_model, target = "Bcell")
panBcell.performance$performance
```

We see that in one of the datasets our simple model did not have an excellent precision (85%).
We can refine it, for instance, by removing potentially contaminating T cell genes (CD3D) and try again:

```{r, fig.width=8,fig.height=12,collapse = T, message = F}
my_scGate_model <- edit_model(model = my_scGate_model, name = "Tcell", signature = c("CD3D"),negative = T)
panBcell.performance <- test_my_model(my_scGate_model, target = "Bcell")
panBcell.performance$performance

```
Now performance is very high for the three datasets.

## Using our repository of pre-defined gating models

We also provide some pre-defined gating models, these can be retrieved with get_scGateDB()

Load default models from DB
```{r,collapse=TRUE}
models.DB <- scGate::get_scGateDB(force_update = T)

names(models.DB$human$generic)

names(models.DB$mouse$generic)

```

### Have a look at some model (PanBcell for B cells and plasma cells)

To visualize complex models in tree-like structures, we provide plot_tree() function
This is a model to purify plasma (B) cells with high accuracy

```{r message=F,results=F}
my_scGate_model <- models.DB$human$generic$Plasma_cell
plt.tree <- scGate::plot_tree(my_scGate_model)
plt.tree
```


Run an example, now on another tumor dataset by Zilionis et al.

```{r}
obj <- testing.datasets[["Zilionis"]]
DimPlot(obj, label = T, repel = T, group.by = "cell_type") + theme(legend.position = "none", aspect.ratio = 1) 
```


```{r, collapse=T}
# run scGate
obj <- scGate(obj, model = my_scGate_model)
DimPlot(obj, cols = c(list(Impure = "gray", Pure = "green"))) + theme(aspect.ratio = 1) 
```

Visualizing filtering results by level. We can see how the purity increases at each level.

```{r,fig.height=8,fig.width=10 }
plots <- scGate::plot_levels(obj)
plots[["annotation"]] <- DimPlot(obj, group.by = "cell_type",label = T, repel =T,label.size = 3) + ggtitle("Author's manual annot") + NoLegend() 
wrap_plots(plots, ncol = 2)
```

Evaluate performance of filtering compared to original annotation
```{r,collapse=T}
perf.scGate = scGate::performance.metrics(actual = obj$Plasma_cell, pred = obj$is.pure== "Pure")
perf.scGate
```


